Index: av1/common/resize.c
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/av1/common/resize.c b/av1/common/resize.c
--- a/av1/common/resize.c	(revision 87460cef80fb03def7d97df1b47bad5432e5e2e4)
+++ b/av1/common/resize.c	(date 1646539565089)
@@ -1355,6 +1355,9 @@
   if (scaling_required) {
     const int num_planes = av1_num_planes(cm);
     const SequenceHeader *seq_params = cm->seq_params;
+    const bool can_use_optimized = scaled_width * 4 >= unscaled->y_crop_width &&
+                                   scaled_height * 4 >= unscaled->y_crop_height;
+
 
     // Reallocate the frame buffer based on the target dimensions when scaling
     // is required.
@@ -1367,14 +1370,15 @@
                          "Failed to allocate scaled buffer");
 
 #if CONFIG_AV1_HIGHBITDEPTH
-    if (use_optimized_scaler && cm->seq_params->bit_depth == AOM_BITS_8) {
+    if (use_optimized_scaler && can_use_optimized
+        && cm->seq_params->bit_depth == AOM_BITS_8) {
       av1_resize_and_extend_frame(unscaled, scaled, filter, phase, num_planes);
     } else {
       av1_resize_and_extend_frame_nonnormative(
           unscaled, scaled, (int)cm->seq_params->bit_depth, num_planes);
     }
 #else
-    if (use_optimized_scaler) {
+    if (use_optimized_scaler && can_use_optimized) {
       av1_resize_and_extend_frame(unscaled, scaled, filter, phase, num_planes);
     } else {
       av1_resize_and_extend_frame_nonnormative(
